hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/SCMCommonPlacementPolicy.java

```diff
@@ -445,6 +445,7 @@ public ContainerPlacementStatus validateContainerPlacement(
       }
     }
     List<Integer> currentRackCount = new ArrayList<>(dns.stream()
+        .filter(d -> !(d.isDecommissioned()))
```

Comment:
> Again here, I wonder about maintenance replicas. It would be possible for there to be extra copies of maintenance replicas in the cluster, like with decommission, and until the node goes offline, there would be too many replicas.

Are you saying that we should occasionally filter maintenance replicas as well? We skip them, since we expect a maintenance node to not be stopped.

> But as the container is effectively over-replicated, we have to adjust the max nodes per rack upwards by the delta between the actual replica count and expected replicas.
>
> So if we add to that max per rack Math.max(0, dns.size() - replicas) = 1, then we get 3 allowed per rack, which means the policy would not be violated any more.

I don't understand this part, can you explain it more? Where can I find it?

In your example,  `(3xin_service 1xdecommissioning)` the container isn't over-replicated for SCM. The replica on the decommissioned node isn't counted. But the placement policy will say that the container is mis-replicated. As long as the SCM considers the container properly replicated, it makes sense to me to filter any replicas on offline nodes, in the policy as well.

Reply:
What I am saying is that we should not filter nodes here, but adjust the max-nodes-per-rack higher if there are more replicas passed in. There is the decommission scenario, maintenance, plus there are scenarios in replication manager where we maintain extra copies of unhealthy replicas too, so you could get 3 x in_service plus 1 or more unhealthy.

If you have 3 replicas, then the expectation is 2 racks and a max of 2 per rack. But what if there are 4 replicas for rep-factor 3. Then you need 2 or 3 racks. Plus at most 3 per rack.

So if you adjust the max per rack higher based on the number of replicas available, then I think it would fix the problem.

Reply:
> So if you adjust the max per rack higher based on the number of replicas available, then I think it would fix the problem.

ok, I get it. Let me look into that. Thanks!

Reply:
@sodonnel I was thinking that we need to stop reporting mis-replication in case of excessive replicas due to offline nodes. Are you saying that we should include unhealthy replicas as well? Stop reporting mis-replication if we have proper replication and the excessive number of replicas are unavailable, either due to offline nodes or being unhealthy?

Reply:
@sodonnel I've been looking into your suggestion. With that change, we will be ignoring mis-replication for every over-replicated container.

Let's say we have this scenario
* RackAware policy
* 7 datanodes, 3 are IN_SERVICE, 4 are offline
* 2 of the offline nodes get back IN_SERVICE
* Now we have, 5 nodes IN_SERVICE, 2 offline
* 5 available replicas and we only need 3, so the container is over-replicated

Initially
```
/rack0/node0  -> in_service       - used
/rack0/node1  -> in_service       - used
/rack0/node2  -> offline
/rack0/node3  -> offline
/rack0/node4  -> offline
/rack1/node5  -> in_service       - used
/rack1/node6  -> offline
```

After 2 offline coming back
```
/rack0/node0  -> in_service                     - used
/rack0/node1  -> in_service                     - used
/rack0/node2  -> offline ->  in_service         - used
/rack0/node3  -> offline
/rack0/node4  -> offline
/rack1/node5  -> in_service                     - used
/rack1/node6  -> offline -> in_service          - used
```
The placement will be [3, 2] where we need [2, 1]. The container is mis-replicated but because we adjust the maxReplica number by adding the replicaDelta, we get that the policy is satisfied. 

`maxReplicaPerRack = 2` 
`dns.size() = 7` 
`replicas = 3`
`Math.max(0, dns.size() - replicas) = Math.max(0, 4) = 4` regardless of whether the excessive replicas are healthy or not.
`maxReplicaPerRack = 6`

Even if we add a check to increase the maxReplica number, only if there are decommission or maintenance nodes, it won't make a difference. We can't distinguish between replicas on offline nodes and replicas on online nodes.

This approach will effectively ignore `mis-replication` when caused by `over-replication`.

Reply:
The new replication manager does not report on mis-replication when a container is over or under-replicated. It expects those conditions to be fixed first.

We have to consider what does mis-replication really mean - the original intention was to indicate that the rack tolerence of the container isn't high enough. If you have enough racks, but too many nodes on some racks due to over-replication, then the container isn't really mis-replicated in my opinion.

I think that one reason the max-per-rack was brought it, was for EC + Rack Scatter. Say you have 3 racks and EC 3-2. The idea distribution is 2, 2, 1. However 3, 1, 1 is also possible, but its not good, as if you lose the rack with 3, the EC data is no longer readable. In that case, the container would be rightly mis-replicated.

When you have over-replication, I think it is fine to report the container as over-replicated, and so long as the minimum racks is met, then it is not mis-replicated. After over-replication has been fixed, then mis-replication can come into play.

Reply:
@sodonnel Thanks for the explanation. I agree. I'll add all that as a comment because other people might have the same questions and confusion.

Since over-replication shouldn't be considered as mis-replication, we should adjust the max replicas per node every time there are excessive replicas, right? We don't need to check for decommission or maintenance to perform the operation.

I'm referring to something like this
```java
    int maxReplicasPerRack = getMaxReplicasPerRack(replicas,
            Math.min(requiredRacks, numRacks));

    boolean offlineNodeReplicas = dns.stream()
        .anyMatch(d -> d.isMaintenance() || d.isDecommissioned());

    if (offlineNodeReplicas) {
      // Adjust max replicas per rack for excessive replicas belonging to offline nodes.
      maxReplicasPerRack += Math.max(0, dns.size() - replicas);
    }
```

Reply:
ok, thanks. I'll add some tests and push the changes.

Reply:
@sodonnel Since we are allowing over-replication, I think we should also extend the check in `ContainerPlacementStatusDefault#isPolicySatisfied()` to something like this

```java
cnt <= maxReplicasPerRack && maxReplicasPerRack - cnt <= 1);
```
so that we can report mis-replication if there is too much uneven distribution.

In the above example, 

> I think that one reason the max-per-rack was brought it, was for EC + Rack Scatter. Say you have 3 racks and EC 3-2. The idea distribution is 2, 2, 1. However 3, 1, 1 is also possible,

3, 1, 1 won't be reported as mis-replication with the new code addition. But if we check that the replica difference between racks isn't more than 1, we can identify such cases.

What do you think?

Reply:
3, 1, 1 should still be reported as mis-replicated for EC-3-2, as there should be 5 replicas. There is no excess / over replication in this example.

If we had 4, 1, 1 - this has an excess of 1. It also has an idea max-per-rack of 2, but we adjust it 1 higher due to the excess to 3, and hence it should be reported as mis-replicated. However 3, 2, 1 would not be.